src/main/java/io/stargate/sgv2/jsonapi/service/operation/reranking/RerankingTask.java

```diff
@@ -149,26 +149,33 @@ float[] sortVector() {
     return sortVector;
   }
 
+  public record DeferredCommandWithSource(
+      Rank.RankSource rankSource, DeferredCommandResult deferredRead) {}
+
   private record DeduplicationResult(
       int totalReads,
       int totalDocuments,
       int droppedDocuments,
       List<ScoredDocument> deduplicatedDocuments) {}
 
-  private DeduplicationResult deduplicateResults(List<CommandResult> rawReadResults) {
+  private DeduplicationResult deduplicateResults() {
 
-    if (rawReadResults.isEmpty()) {
-      throw new IllegalArgumentException("rawReadResults must not be empty");
+    if (deferredReads.isEmpty()) {
+      throw new IllegalArgumentException("deferredReads must not be empty");
     }
 
     // This code relies on working with collections where the documents all have _id
     // will need to change for tables and rows
-    // Keyed on the _id as a JsonNode
-
     ScoredDocumentMerger merger = null;
     int totalReads = 0;
 
-    for (var commandResult : rawReadResults) {
+    for (var deferredAndSource : deferredReads) {
+      var commandResult =
+          Objects.requireNonNull(
+              deferredAndSource.deferredRead().commandResult(),
```

Comment:
Not the biggest problem in the world but I think this will eagerly construct error String that is almost never needed.

Reply:
changed to 

```java
      var commandResult = deferredAndSource.deferredRead().commandResult();
      if (commandResult == null){
        throw new IllegalStateException("Deferred read from source %s returned null commandResult"
                  .formatted(deferredAndSource.rankSource()));
      }
```